servlet/src/main/java/io/grpc/servlet/ServletServerStream.java

```diff
@@ -0,0 +1,425 @@
+            if (logger.isLoggable(FINEST)) {
+              logger.log(
+                  FINEST,
+                  "[{0}] outbound data: length = {1}, bytes = {2}",
+                  new Object[]{
+                      logId, numBytes, toHexString(byteBuffer.bytes, byteBuffer.readableBytes())});
+            }
+          }
+          if (!outputStream.isReady()) {
+            while (true) {
+              if (writeState.compareAndSet(curState, curState.withStillWritePossible(false))) {
+                return;
+              }
+              curState = writeState.get();
+            }
+
+          }
+        } catch (IOException ioe) {
+          logger.log(WARNING, String.format("[{%s}] Exception writing message", logId), ioe);
+          cancel(Status.fromThrowable(ioe));
+        }
+      } else {
+        writeChain.offer(byteBuffer);
+        if (!writeState.compareAndSet(curState, curState.newState())) {
+          // state changed by another thread, need to check if stillWritePossible again
+          if (writeState.get().stillWritePossible) {
+            ByteArrayWritableBuffer bf = writeChain.poll();
+            if (bf != null) {
+              assert bf == byteBuffer;
+              writeFrame(bf);
+            }
+          }
+        }
+      }
+    }
+
+    @Override
+    public void writeTrailers(Metadata trailers, boolean headersSent, Status status) {
+      if (logger.isLoggable(FINE)) {
+        logger.log(
+            FINE,
+            "[{0}] writeTrailers: {1}, headersSent = {2}, status = {3}",
+            new Object[] {logId, trailers, headersSent, status});
+      }
+      if (!headersSent) {
+        // Discard any application supplied duplicates of the reserved headers
+        trailers.discardAll(CONTENT_TYPE_KEY);
+        trailers.discardAll(GrpcUtil.TE_HEADER);
+        trailers.discardAll(GrpcUtil.USER_AGENT_KEY);
+
+        resp.setStatus(HttpServletResponse.SC_OK);
+        resp.setContentType(CONTENT_TYPE_GRPC);
+
+        byte[][] serializedHeaders = TransportFrameUtil.toHttp2Headers(trailers);
+        for (int i = 0; i < serializedHeaders.length; i += 2) {
+          resp.addHeader(
+              new String(serializedHeaders[i], StandardCharsets.US_ASCII),
+              new String(serializedHeaders[i + 1], StandardCharsets.US_ASCII));
+        }
+      } else {
+        byte[][] serializedHeaders = TransportFrameUtil.toHttp2Headers(trailers);
+        for (int i = 0; i < serializedHeaders.length; i += 2) {
+          String key = new String(serializedHeaders[i], StandardCharsets.US_ASCII);
+          String newValue = new String(serializedHeaders[i + 1], StandardCharsets.US_ASCII);
+          String value = trailerSupplier.get().putIfAbsent(key, newValue);
+          if (value != null) {
+            trailerSupplier.get().put(key, value + "," + newValue);
+          }
+        }
+      }
+
+      while (true) {
```

Comment:
How does the servlet know when we are done writing? I see we have placed the trailers in place ready to be read, but what is the trigger? I sort of expected to see an `outputStream.close()` or similar.